sasmodels/models/micromagnetic_FF_3D.c

```diff
@@ -0,0 +1,196 @@
+  return f;
+}
+
+static double fqMyreal( double x, double y, double z, double Mz, double Hkx, double Hky, double Hi, double Ms, double A, double D)
+{
+  const double q=MAG_VEC(x, y, z);  
+  const double f = reduced_field(q, Ms, Hi, A)*(Hky*(1.0+reduced_field(q, Ms, Hi, A)*x*x/q/q)-Ms*Mz*y*z/q/q*(1.0+reduced_field(q, Ms, Hi, A)*DMI_length(Ms, D,q)*DMI_length(Ms, D,q))-Hkx*reduced_field(q, Ms, Hi, A)*x*y/q/q)/(1.0+reduced_field(q, Ms, Hi, A)*(x*x+y*y)/q/q -square(reduced_field(q, Ms, Hi, A)*DMI_length(Ms, D,z)));
+  return f;
+}
+
+static double fqMyimag( double x, double y, double z, double Mz, double Hkx, double Hky, double Hi, double Ms, double A, double D)
+{
+  const double q=MAG_VEC(x, y, z);  
+  const double f = reduced_field(q, Ms, Hi, A)*(Ms*Mz*(1.0+reduced_field(q, Ms, Hi, A))*DMI_length(Ms, D,x)-Hkx*reduced_field(q, Ms, Hi, A)*DMI_length(Ms, D,z))/(1.0+reduced_field(q, Ms, Hi, A)*(x*x+y*y)/q/q -square(reduced_field(q, Ms, Hi, A)*DMI_length(Ms, D,z)));
+  return f;
+}
+
+static double
+Iqxy(double qx, double qy, double nuc_radius, double nuc_thickness, double mag_radius, double mag_thickness, double hk_radius, double nuc_sld_core, double nuc_sld_shell, double nuc_sld_solvent, double mag_sld_core, double mag_sld_shell, double mag_sld_solvent, double hk_sld_core, double Hi, double Ms, double A, double D,  double up_i, double up_f, double alpha, double beta)
+{
+  const double q=MAG_VEC(qx, qy, 0);    
+  if (q > 1.0e-16 ) {
+    const double cos_theta=qx/q;
+    const double sin_theta=qy/q;
+
+    double qrot[3];
+    set_scatvec(qrot,q,cos_theta, sin_theta, alpha, beta);
+    // 0=dd.real, 1=dd.imag, 2=uu.real, 3=uu.imag,  4=du.real, 6=du.imag,  7=ud.real, 5=ud.imag
+    double weights[8];
+    set_weights(up_i, up_f, weights);
+   
+    double mz=fq(q, mag_radius, mag_thickness, mag_sld_core, mag_sld_shell, mag_sld_solvent);
+    double nuc=fq(q, nuc_radius, nuc_thickness, nuc_sld_core, nuc_sld_shell, nuc_sld_solvent);
+	
+
+    double cos_gamma, sin_gamma;
+    double sld[8];
+    //loop over random anisotropy axis with isotropic orientation gamma for Hkx and Hky
+    //To be modified for textured material see also Weissmueller et al. PRB 63, 214414 (2001)
+    double total_F2 = 0.0;
+    for (int i=0; i<GAUSS_N ;i++) {
+      const double gamma = M_PI * (GAUSS_Z[i] + 1.0); // 0 .. 2 pi
+      SINCOS(gamma, sin_gamma, cos_gamma);	
+      //Only the core of the defect/particle in the matrix has an effective
+      //anisotropy (for simplicity), for the effect of different, more complex
+      // spatial profile of the anisotropy see Michels PRB 82, 024433 (2010)
+      double Hkx= fq(q, hk_radius, 0, hk_sld_core, 0, 0)*sin_gamma;
+      double Hky= fq(q, hk_radius, 0, hk_sld_core, 0, 0)*cos_gamma;
+	  
```

Comment:
Lots of shared subexpressions in fqMxreal, etc.; performance improvements for inlining the calculations and sharing code.